### ISSUE TITLE
Provide macros which perform formatting after calling `gettext`

### ISSUE THREAD

**Issue by Minoru** [Koka/gettext-rs#86]
We have macros that do something like `format!(gettext(…), …)`, but at runtime (`format!` expects a string literal, so we can't use that). Our macros supported `{}` and `{n}`, and there was [a PR](https://github.com/Koka/gettext-rs/pull/83) to add escaping (`{{` and `}}`). That PR uncovered a problem: if the format string is invalid, what should our macros return?

To find out what user's expectations are, [I looked at how people use gettext-rs functions](https://github.com/Koka/gettext-rs/pull/83#discussion_r706646772):

* GTK apps (mentioned in #18) use Rust-style formats and assume that the string is correct (i.e. they don't panic and they don't return Result)
* in my own app, Newsboat, I wrap `strnprintf` C function, use printf-style formats, and assume the format string is correct
* gstore crate [has its own macro](https://gitlab.com/loers/gstore/-/blob/5e367356a7db2254635241224d238fdb0d09570f/src/l10n.rs#L131). It uses Rust-style formats, and assumes the format string is correct
* [tr crate](https://crates.io/crates/tr). Rust-style formats, assumes correctness
* [serde_yaml crate](https://github.com/cecton/serde-gettext) uses Rust-style formats, returns `Result`

I also looked at how others solve this problem:

* Python's `format` method throws `ValueError` if the format string contains imbalanced curly braces
* C++ `std::vformat` [throws](https://en.cppreference.com/w/cpp/utility/format/vformat) too
* [runtime-fmt](https://crates.io/crates/runtime-fmt) and [dynfmt](https://crates.io/crates/dynfmt) crates return `Result`

So it looks like the expected solution would be returning `Result`. [Some ideas](https://github.com/Koka/gettext-rs/pull/83#discussion_r706823824) from @delight-aug:

> From the perspective of the user of macros, with Result, I could call auto-bug-reporting and display a message to users, that something went wrong. Or I could just unwrap it all the time and let my program crash because of some string. Also, it doesn't look good when you want to include a phrase in GtkLabel, but you need to match it first (or unwrap it in place). Maybe a good decision is to let users specify an action to run on a broken string? So they set up a closure near textdomain bindings. And by default just swallow things...

This, plus the lack of clear winner among the "runtime formattting" crates, makes me think we should postpone this work, and gather more ideas of what users actually need.

If you'd use a macro that does something like `format!(gettext(…), …)`, please comment below and explain:
- what format specifiers you expect to use. Just plain `{}`? Positional `{1}`? Named `{foo}`? How about complex ones like `{:3.4x}`?
- if the translated string is not a valid format sequence (e.g. the translator made a typo), what do you expect the macro to do? Panic? Communicate the failure via `Result`? Return an empty string? Run some closure that you set up earlier?

**Comment by delight-aug** [comment]
I think the first thing to do if translated string has broken formatting is to use English (default) string. But what to do if its formating is broken, too?

**Comment by zecakeh** [comment]
I'm working on [Fractal](https://gitlab.gnome.org/GNOME/fractal), a GTK app.

From my point of view, a sane fallback using the default string seems like a good default, although providing a custom closure would be a more generic solution.

> But what to do if its formating is broken, too?

The macro should probably check that the formatting of the string is good and fail `cargo check` if it doesn't.

From a translator point of view, I believe that named format specifiers are a must have. A specifier with a good name is easier to understand. Also, the translator might need to change the variables' order and it's not doable without it. I do not think that complex specifiers are a must-have.

We are rewriting our app from scratch because of several big changes we needed to make. Our legacy code uses [custom i18n methods](https://gitlab.gnome.org/GNOME/fractal/-/raw/master/fractal-gtk/src/util/i18n.rs) based on an old version of this crate, to allow us to use named variables. They might be of help. Of course the syntax is not as friendly as a macro would be, and we'd like to avoid having a custom implementation.

**Comment by delight-aug** [comment]
> The macro should probably check that the formatting of the string is good and fail cargo check if it doesn't.

That's a good idea we missed. This crate is just a bindings library for gettext, so I thought we have to format the `String` that actual gettext returns at runtime. But, I think idealistically nothing stops us from using the str provided to macro at compile-time to verify that the default string isn't broken (to at least have a working fallback string).

That being said, I'm struggling to think about how to implement it better. Just in case, Is there anything that allows to write custom compile-time checks for macros?

**Comment by zecakeh** [comment]
> That being said, I'm struggling to think about how to implement it better. Just in case, Is there anything that allows writing custom compile-time checks for macros?

With procedural macros, if it panics, it gives a compile-time error. So you can just use `panic!("Error message")` for custom errors.

**Comment by delight-aug** [comment]
I have looked at procedural macros, and it seems absolutely possible to create the type of macros you suggested. I'll start working on it after 9th December, the deadline for my coursework. However, I can't say whether it will be implemented in this crate — that's up to @Minoru to decide.

**Comment by bhaible** [comment]
The next version of GNU gettext will have Rust support. It will recommend to use the `formatx` package. Here's the sample code:
```
/* gettext-rs
   Home: https://crates.io/crates/gettext-rs
         https://lib.rs/crates/gettext-rs
         https://github.com/gettext-rs/gettext-rs
   Documentation: https://docs.rs/gettext-rs/latest/gettextrs/ */
use gettextrs::*;

/* formatx
   Home: https://crates.io/crates/formatx
         https://lib.rs/crates/formatx
         https://github.com/clitic/formatx
   Documentation: https://docs.rs/formatx/latest/formatx/ */
use formatx::*;

use std::process;

fn main() {
  setlocale (LocaleCategory::LcAll, "");

  textdomain ("hello-rust").expect ("textdomain failed");
  bindtextdomain ("hello-rust", "@localedir@").expect ("bindtextdomain failed");
  /* gettext-rs requires UTF-8 encoding.  */
  bind_textdomain_codeset ("hello-rust", "UTF-8").expect ("bind_textdomain_codeset failed");

  println! ("{}", gettext ("Hello, world!"));
  println! ("{}", formatx! (gettext ("This program is running as process number {}."), process::id ()).unwrap ());
}
```

With this, I believe, you can close this ticket.

**Comment by Minoru** [comment]
Yes, formatx is an option (just like runtime-fmt and dynfmt are), but I still think we can do better: a custom macro could recover from formatting error by formatting the original string instead.

**Comment by bhaible** [comment]
> formatx is an option (just like runtime-fmt and dynfmt are)

https://www.reddit.com/r/rust/comments/13p3nnn/library_for_formatting_userprovided_format_strings/
says that `runtime_fmt` "no longer compiles". I confirm: Simply adding
```
runtime-fmt = "0.4.1"
```
to Cargo.toml produces compilation errors:
```
error: could not compile `runtime-fmt` (lib) due to 13 previous errors
```

And `dynfmt` suffers from the same problem as `strfmt`: It does not support formatting where some of the values are strings and some of the values are numbers (without an excessive amount of contortions): You can't write `string.format(&["cheese", n])` where `n` is a number.

In messages to the user, it is completely normal to have both strings and numbers (e.g. a file name and a line number). `formatx` is the only crate that fulfils this requirement.

> I still think we can do better: a custom macro could recover from formatting error by formatting the original string instead.

This is not needed, because `msgfmt -c` verifies that the translation of the format string and the untranslated format string are compatible. Just like for other programming languages. Given this, there's no need to merge the two distinct processing steps (lookup of the translation, then formatting) into a single step.

**Comment by Minoru** [comment]
Oh, I knew none of that. Thanks for educating me! You're right, let's close the issue then.
